### PR TITLE
Run lint/typechecking in latest python + update pre-commit

### PR DESCRIPTION
**Pull request by cooperlees**
- Lets move to latest and greatest of lints
- Update mypy + flake8

### CODE REVIEW

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -18,8 +18,6 @@ jobs:
 
       - name: Set up Python
         uses: actions/setup-python@v2
-        with:
-          python-version: 3.7
```
Comment (JelleZijlstra) [inline .github/workflows/lint.yml:22]: Would it be useful to specify the version here so things don't break randomly in the future when a new version is released?

Comment (ichard26) [reply .github/workflows/lint.yml:22]: I think the rational was that making the version unbounded reduces the amount of work we have to do when a new Python version rolls around. Although I feel like we are trading that for the (very annoying although probably uncommon) situations when our CI break randomly. Personally still not fully convinced, but Cooper is the upgrade early and then fix the resulting issues type of guy from what I've heard.

Comment (cooperlees) [reply .github/workflows/lint.yml:22]: Exactly. I hate keeping things up to date.

Comment (JelleZijlstra) [reply .github/workflows/lint.yml:22]: OK, let's do it

**Pull request merged**